# GMI-PAN
Bacterial plasmid annotation pipeline

The emergence of antibiotic resistance is a natural occurrence in microorganisms, but its prevalence in clinically significant strains has been accelerated due to the excessive use of antibiotics in humans and farm animals. The indiscriminate use of antibiotics has been identified as a key factor in the development of multidrug resistance, driven by selective pressure from an ecological perspective.

In this repository, i developed a bioinformatics tool called "GMI-PAN", under the code "Contannotate", for the evaluation and annotation of antibiotic and metal resistance profiles, virulence factors, and mobile genetic elements in assembled nucleotide sequences of prokaryotic nature, optimized for the Enterobacteriaceae family. This tool, better known as a wrapper, allows for the automation of an annotation pipeline that integrates various previously developed tools, significantly reducing the time required for work.

This is still a software in an early phase of development and i'm taking my first dive into open-software/hardware so please be patient with my lack of etiquette. Files appended here are products and evidence of a first functional iteration of GMI-PAN that is part of my thesis to obtain the professional academic grade in the Molecular Biotechnology Engineer career.

GMI-PAN is based in annotation modules such as PROKKA, PlasmidVerify, PlasmidFinder, RGI, PhageBoost, NHMMER and IVIP. Thanks to the respective or respectives authors for their contribution and please visit their work!

PROKKA: https://github.com/tseemann/prokka

RGI: https://github.com/arpcard/rgi/releases

I-VIP: https://github.com/caozhichongchong/I-VIP

PlasmidFinder: https://bitbucket.org/genomicepidemiology/plasmidfinder/src/master/

PlasmidVerify: https://github.com/ablab/plasmidVerify

PhageBoost: https://github.com/ku-cbd/PhageBoost

NHMMER: https://github.com/EddyRivasLab/hmmer
